### ISSUE TITLE
Fish prints error on startup

### ISSUE THREAD

**Issue by wucke13** [NixOS/nixpkgs#30645]
## Issue description
On startup, fish prints `set: Tried to change the read-only variable “_”` both in my terminal emulator and the ttys ~~, but not in the terminal emulator when the Xsession is started via slim. On lightdm or manual start through startx however the error persists.~~

### Steps to reproduce
Setup fish as default shell via
`programs.fish.enable = true ;` and `users.extraUsers.<username>.shell = pkgs.fish;`

Start a terminal emulator with fish as shell.

## Technical details

* System: (NixOS: `17.09.1756.c99239bca0 (Hummingbird)`)
* Nix version: (run `nix-env (Nix) 1.11.15`)
* Nixpkgs version: (run `"17.09.1756.c99239bca0"`)
* Sandboxing enabled: (run `build-use-sandbox = false`)

**Comment by wucke13** [comment]
I found some more Information; here is some context with high debug level:
```
<4> fish: io_buffer_t::read: blocking read on fd 3
<4> fish: Exec job 'set value (echo $variable | /nix/store/d8xpyjpammvl3839m4drsx1xk7m2d91q-gnused
-4.4/bin/sed 's/[^=]*=//')' with id 10
<3> fish: Skipping fork: no output for internal builtin 'set'
<3> fish: Set status of set value (echo $variable | /nix/store/d8xpyjpammvl3839m4drsx1xk7m2d91q-gn
used-4.4/bin/sed 's/[^=]*=//') to 0 using short circuit
<3> fish: Job is constructed
<4> fish: Continue job 10, gid -2 (set value (echo $variable | /nix/store/d8xpyjpammvl3839m4drsx1x
k7m2d91q-gnused-4.4/bin/sed 's/[^=]*=//')), COMPLETED, NON-INTERACTIVE
<5> fish: intern /nix/store/y7rdwg8vp8m9snbyy0mcixxbd3zpz1xj-fish-foreign-env-git-20170324/share/f
ish-foreign-env/functions/fenv.apply.fish
<4> fish: Exec job 'test "$key" = 'PATH'' with id 9
<3> fish: Skipping fork: no output for internal builtin 'test'
<3> fish: Set status of test "$key" = 'PATH' to 1 using short circuit
<3> fish: Job is constructed
<4> fish: Continue job 9, gid -2 (test "$key" = 'PATH'), COMPLETED, NON-INTERACTIVE
<4> fish: Exec job 'set -g -x $key $value' with id 11
<3> fish: Skipping fork: ordinary output for internal builtin 'set'
set: Tried to change the read-only variable “_”
<3> fish: Set status of set -g -x $key $value to 1 using short circuit
<3> fish: Job is constructed
<4> fish: Continue job 11, gid -2 (set -g -x $key $value), COMPLETED, NON-INTERACTIVE
<3> fish: Job is constructed
```
Looks like this is happening from the dynamical build environment for fish. This explains the special behavior of described in the original issue, my `startx` session runs inside of an fish session, so the environment is not build again.

**Comment by therealpxc** [comment]
This is due to differences between the ways that `fish` and POSIX shells work, exposed by the wrapping we do with the `fenv` tool, which allows us to load configurations designed for POSIX shells (like `bash`, for example) into a `fish` environment, despite the syntactical incompatibilities between the languages. In this case, the error message is reporting a totally harmless issue: `_` is used to represent the name of the running process in a shell, and `fish` does not allow this variable to be manipulated manually. Because it gets changed in the `bash` environment in the course of the `fenv` call, `fenv` tries to change it in `fish` as well. Because `fish` does not allow this, it prints a warning.

There are some other messages [we've suppressed](https://github.com/NixOS/nixpkgs/blob/master/pkgs/shells/fish-foreign-env/hide-path-warnings.patch), and we could take a similar approach here. I was thinking that maybe we can just have `fenv`check for the variable with that name and skip exporting it. What do you think, @rnhmjoj @cbarrett ?

**Comment by cbarrett** [comment]
That sounds right to me @therealpxc.

**Comment by rnhmjoj** [comment]
@therealpxc Sure. Something like this should do:

```fish
set ignore PATH _

if test "$key" = 'PATH'
  set value (echo $value | tr ':' '\n')
end

if contains $key $ignore
  set -g -x $key $value ^/dev/null
else
  set -g -x $key $value
end
```

While we are at it, are there any other warnings?

**Comment by therealpxc** [comment]
I was thinking that in the case of `$_`, we might as well not attempt to set, rather than doing so and suppressing the error.

I'm not aware of any other bad startup messages. Are you, @wucke13?

**Comment by wucke13** [comment]
No, that was the only error message I got on startup. Looks like a reasonable fix to just suppress the message, as there aren't any follow up errors on this AFAIK.

**Comment by yrashk** [comment]
I opened a PR  with @rnhmjoj's patch: https://github.com/NixOS/nixpkgs/pull/31646

Help with testing will be highly appreciated.

**Comment by rnhmjoj** [comment]
Can we close it?

**Comment by yrashk** [comment]
On my end, #31646 did the trick.

**Comment by wucke13** [comment]
On my end, there is no error left too.

**Comment by bbc2** [comment]
When I SSH into my recently installed NixOS, I get:

```
set: Tried to change the read-only variable “_”
Welcome to fish, the friendly interactive shell
```

Could this issue have been reintroduced?

Subsequent invocations of fish are fine (like with `fish` or `sudo -s`).

## Technical details

 - system: `"x86_64-linux"`
 - host os: `Linux 4.9.84, NixOS, 17.09.3094.211309626af (Hummingbird)`
 - multi-user?: `yes`
 - sandbox: `no`
 - version: `nix-env (Nix) 1.11.16`
 - channels(root): `"nixos-17.09.3094.211309626af"`
 - nixpkgs: `/nix/var/nix/profiles/per-user/root/channels/nixos/nixpkgs`

**Comment by florianjacob** [comment]
This was not backported to 17.09 I think, the fix will be in 18.03.

**Comment by bbc2** [comment]
Indeed, installing the unstable version of fish-foreign-env solved it.  Thanks for the clarification.
